### ISSUE TITLE
Generate structural data for each Drug - Target complex by using PLIP tool

### ISSUE THREAD

**Issue by MarineGirardey** [opentargets/issues#1979]
Next step of my internship.

# Background

- With Molecules/Drugs from parquet files, the PDB identifiers have been obtained from Inchikeys. 
- Structure PDB identifiers of the Drug - Target complex has been gotten from the PDBe API. 
- Targets from an ensembl csv file has been extracted thanks to the PDB structure id.
- Finally Drugs has been filtered to keep the relevant.

# Tasks

Now the next task is to generate the structural data about each Drug - Target to have access to the interacting residues and all needed details about the interactions. Also generate a 3D visualisation of the complex.
The [PLIP](https://github.com/pharmai/plip) python module is the tool used for this task.

# To summarize the tasks are:

- [x]  Write a script to generate the data with PLIP for a small set of data

- [x]  Test the script for the whole set of data

- [x] Improve the script

**Comment by MarineGirardey** [comment]
# Generate PLIP output by parallelize the process with Dask
**Reproducable error**
At the end you should have this error:

```
BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```

*Contact me if there is a bug (other then above lol)*

First of all
```bash
conda create plip_env
conda create --name plip_env
conda activate plip_env
conda install python=3.8.12
conda install matplotlib=3.5.1
conda install pandas=1.4.1
conda install mamba -n base -c conda-forge
mamba create -n opencadd opencadd
conda install -c conda-forge plip
conda install pyspark=3.2.1
```

Let's begin
```python
from plip.structure.preparation import PDBComplex
from plip.exchange.report import BindingSiteReport
from pyspark.sql import SparkSession
from plip.basic import config
import pandas as pd
import dask.dataframe as dd
from pyspark.sql.functions import udf, col, collect_set, pandas_udf
```

Imports
```python
spark = (
    SparkSession.builder
    .getOrCreate()
)
```

Small change to avoid errors for certain structure (needed)
```python
config.DNARECEPTOR = True
```

Load the csv file with structure in a DataFrame
```python
compound_structure_df = (
    spark.read
        .csv(structure_for_plip_path, sep=',', header=True, comment='#')
)
```

Function which generate the plip output (simplified) which need to be parallelise
```python
def run_plip(structure, drug):
    """This function fetches the pdb file from ePDB server as a string

    Args:
        valid_types (str)
        plip_wanted_columns
        combination_dict
    Returns:
        A PySpark dataframe
    """

    test_list = []

    protlig = PDBComplex()
    
    try:
        protlig.load_pdb(f'output_files/pdb_structure_files/pdb{structure}.ent')  # load the pdb file
        test_list.append(f"{structure} loaded")

    except FileNotFoundError:
        test_list.append(f"{structure} not loaded")
        pass

    return test_list
```

Group the drugs together for a same structure
```python
modified_df = (compound_structure_df
               .groupby("pdbStructureId")
               .agg(collect_set(col("pdbCompoundId")).alias("pdbCompoundIdAgg"))
               )

modified_df.show()
```

Set dask
```python
modified_pd_df = modified_df.toPandas()
partition = dd.from_pandas(modified_pd_df, npartitions=10)
partition
```

Complicated thing that I've found [here](https://stackoverflow.com/questions/45545110/make-pandas-dataframe-apply-use-all-cores)
```python
res_df = (partition
       .assign(new_col = partition.map_partitions(lambda df: df
                       .apply(
                              lambda row: run_plip(*row), axis=1
                              ), meta=(None, 'f8')))
       .compute(scheduler='processes')
)
```

```python
res_df
```

**To save pdb files locally**

Transform the DataFrame into a dict (can't use any udf here)
```python
combinations = {}
for i in compound_structure_pd_df["pdbStructureId"].tolist():
    combinations[i] = compound_structure_pd_df[compound_structure_pd_df["pdbStructureId"]==i]["pdbCompoundId"].tolist()

combinations
```

Save the local file for each 1934 structure in your computer
```
for structure, drugs in combinations.items():

    pdb_file = fetch_pdb_dl_local(structure, session)
```

The function to get pdb from API (change the path!!!!!)
```python
def fetch_pdb_dl_local(pdb_struct_id, session):
    """This function fetches the pdb file from ePDB server as a string

    Args:
        pdb_struct_id (str)
    Returns:
        structure data in pdb format as string
    """

    url = f'https://www.ebi.ac.uk/pdbe/entry-files/download/pdb{pdb_struct_id}.ent'
    headers={'Content-Type': 'text/plain'}
    response = session.get(url, headers=headers)

    try:
        if response.headers['Content-Type'] != 'text/plain; charset=UTF-8':
            pass

        else:
            open(f'output_files/pdb_structure_files/pdb{pdb_struct_id}.ent', 'wb').write(response.content)

    except JSONDecodeError:
        print(f'Failed to return structures to: {pdb_struct_id}')
        if len(response.json()) == 0:
            pass

    except KeyError:
        print(f'Empty data was returned for: {pdb_struct_id}')
        pass

    except ConnectionError:
        pass
```

**The structure-drug file**

[structure_for_plip_small_set.csv](https://github.com/opentargets/platform/files/8306249/structure_for_plip_small_set.csv)

Good luck

:skull_and_crossbones:
--


<br class="Apple-interchange-newline">

:skull_and_crossbones:
--


<br class="Apple-interchange-newline">

:skull_and_crossbones:
--


<br class="Apple-interchange-newline">

:skull_and_crossbones:
--


<br class="Apple-interchange-newline">

:skull_and_crossbones:
--


<br class="Apple-interchange-newline">

**Comment by DSuveges** [comment]
I think there's something to improve in this function:

```python
    try:
        protlig.load_pdb(f'output_files/pdb_structure_files/pdb{structure}.ent')  # load the pdb file
        test_list.append(f"{structure} loaded")

    except FileNotFoundError:
        test_list.append(f"{structure} not loaded")
        pass
```

You are only catching the `FileNotFoundError` any other exception breaks the process. As the running of the chunks is split by dusk, the error messages might be swallowed and you might not get notification on the actual events. However I managed to run the code on a subset of 1k structures in ~8minutes.

**Comment by MarineGirardey** [comment]
# RESULTS - Generate PLIP output with parallelistion

Parallelisation library : [Pandarallel](https://github.com/nalepae/pandarallel)

**Objective: Extract informations about residues interacting in each Drug-Target complexe from structure PDB files by using PLIP.**

### Data we have - extract

|pdbStructureId|pdbCompoundId|chainId|         geneId|uniprotId|
|--------------|-------------|-------|---------------|---------|
|          1d5m|          ALC|      A|ENSG00000227993|   P01903|
|          1d5m|          ALC|      A|ENSG00000277263|   P01903|
|          1d5m|          ALC|      B|ENSG00000196126|   P01911|
|          1d5m|          ALC|      A|ENSG00000204287|   P01903|
|          1e9b|          ATM|      A|ENSG00000168393|   P23919|
|          1ere|          EST|      F|ENSG00000091831|   P03372|
|          1ere|          EST|      C|ENSG00000091831|   P03372|
|          1ere|          EST|      D|ENSG00000091831|   P03372|

### PLIP output - extract

|pdbStructureId|pdbCompoundId| interactionType | protResidueNumber | protResidueType |chainId|
|--------------|-------------|-------------------|-------------------|-----------------|-------|
|          1e9b|          ATM|         saltbridge|                 97|              ARG|      A|
|          1e9b|          ATM|              hbond|                157|              GLN|      A|
|          1e9b|          ATM|            pistack|                 72|              PHE|      A|
|          1e9b|          ATM|hydroph_interaction|                105|              PHE|      A|
|          1ere|          EST|              hbond|                394|              ARG|      A|
|          1ere|          EST|            pistack|                404|              PHE|      A|
|          1ere|          EST|hydroph_interaction|                387|              LEU|      A|
|          1ere|          EST|hydroph_interaction|                391|              LEU|      A|

### Statistics

|     Data    |    Counts    |
|----------|------------|
| Input Structures | 17 507 |
| Input Drugs | 2 213 |
| Input Structure-Drugs combinations | 21 320 |
| Structures with PLIP output | 13 629 |
| Drugs with PLIP output | 2 118 |
| Structure-Drugs combinations with PLIP output | 15 486 |
| GeneId with PLIP output | 2 767 |
| UniprotId with PLIP output| 2 526 |
| Hbond interaction | ≈44,7% (145 275) |
| Metalcomplex interaction | ≈25,7% (83 475) |
| Hydrophobic interaction | ≈15,8% (51 480) |
| Saltbridge interaction | ≈9,3% (30 338) |
| Pistack interaction | ≈3,2% (10 258) |
| Pication interaction | ≈1,1% (3 633) |
| Halogenbond interaction | ≈0,2% (652) |
